src/libraries/Microsoft.Extensions.DependencyInjection.Specification.Tests/src/ActivatorUtilitiesTests.cs

```diff
@@ -198,27 +198,29 @@ public void TypeActivatorRethrowsOriginalExceptionFromConstructor(CreateInstance
         }
 
         [Theory]
-        [InlineData(typeof(string))]
-        [InlineData(typeof(int))]
-        public void TypeActivatorCreateFactoryDoesNotAllowForAmbiguousConstructorMatches(Type paramType)
```

Comment:
If this is no longer true, then this PR does not fix #46132. I'm not implying it should, but this PR seems to bring braking changes to both `CreateInstance` and `CreateFactory`.

Reply:
Following @tarekgh comments:

> [..] we must consider the purpose of the API. it is creating an instance of object. If the user code giving two ways (i.e. two legit constructors) and then we fail that will look weird too.
> 
> The current behavior can be easily defined as, CreateInstance tries to create instance of the object by finding any legit constructor satisfy the creation using the input parameters.

This PR tries to solve the #46132 issue by choosing the most similar ctor in terms of parameters and supplied arguments, instead of relying on ctor order as it was done previously. I thought this should address #46132. I specifically tried to cover it by adding tests for this case in 74796631.

Given:
```csharp
public class Foo 
{
    public Foo(A a, B b, C c, S s) { ... }
    public Foo(A a, C c, S s) { ... }
}

ActivatorUtilities.CreateInstance(provider, typeof(Foo), new C(), new A());
```

It would always choose the `Foo(A a, C c, S s)` ctor regardless of definition order, as it is closer to the supplied arguments than the other ctor.

Does this makes sense?

Reply:
I described the expected behavior I want to see.
Using (A, B) arguments where (B, A) parameters are expected looks magical and throws type (signature) safety out of the window, IMHO.
It all comes to what we want. And I don't want the snippet to work. I want it to throw to keep everything simple.
Originally we have magical poorly described `CreateInstance` and strict, but well defined `CreateFactory`. I want the former to behave as the later: that way I'll have fewer surprises and hard to debug errors, I'm fine with failing fast during development/deployment though. But I'm only one user.

Reply:
> Although, if we had a way to check if `IServiceProvider` has a parameter type registered

Unfortunately, the lack of this feature is a major pain for DI adaptors/implementations/wrappers/anything in itself.